examples/quickstart/executor_with_thread_hooks.cpp

```diff
@@ -0,0 +1,224 @@
+            F&& f, S const& shape, Future&& predecessor, Ts&&... ts) const
+        {
+            return hpx::parallel::execution::bulk_then_execute(exec_,
+                hook_wrapper<F>{*this, std::forward<F>(f)}, shape,
+                std::forward<Future>(predecessor), std::forward<Ts>(ts)...);
+        }
+
+    private:
+        using thread_hook = hpx::util::function_nonser<void()>;
+
+        BaseExecutor& exec_;
+        thread_hook on_start_;
+        thread_hook on_stop_;
+    };
+
+    template <typename BaseExecutor, typename OnStart, typename OnStop>
+    executor_with_thread_hooks<BaseExecutor> make_executor_with_thread_hooks(
+        BaseExecutor& exec, OnStart&& on_start, OnStop&& on_stop)
+    {
+        return executor_with_thread_hooks<BaseExecutor>(exec,
+            std::forward<OnStart>(on_start), std::forward<OnStop>(on_stop));
+    }
+}    // namespace executor_example
+
+///////////////////////////////////////////////////////////////////////////////
+// simple forwarding implementations of executor traits
+namespace hpx { namespace parallel { namespace execution {
+
+    template <typename BaseExecutor>
+    struct is_one_way_executor<
+        executor_example::executor_with_thread_hooks<BaseExecutor>>
+      : is_one_way_executor<typename std::decay<BaseExecutor>::type>
+    {
+    };
+
+    template <typename BaseExecutor>
+    struct is_never_blocking_one_way_executor<
+        executor_example::executor_with_thread_hooks<BaseExecutor>>
+      : is_never_blocking_one_way_executor<
+            typename std::decay<BaseExecutor>::type>
+    {
+    };
+
+    template <typename BaseExecutor>
+    struct is_two_way_executor<
+        executor_example::executor_with_thread_hooks<BaseExecutor>>
+      : is_two_way_executor<typename std::decay<BaseExecutor>::type>
+    {
+    };
+
+    template <typename BaseExecutor>
+    struct is_bulk_one_way_executor<
+        executor_example::executor_with_thread_hooks<BaseExecutor>>
+      : is_bulk_one_way_executor<typename std::decay<BaseExecutor>::type>
+    {
+    };
+
+    template <typename BaseExecutor>
+    struct is_bulk_two_way_executor<
+        executor_example::executor_with_thread_hooks<BaseExecutor>>
+      : is_bulk_two_way_executor<typename std::decay<BaseExecutor>::type>
+    {
+    };
+}}}    // namespace hpx::parallel::execution
+
+int main()
+{
+    std::vector<double> v(1000);
+    std::iota(v.begin(), v.end(), 0.0);
+
+    std::atomic<std::size_t> starts(0);
+    std::atomic<std::size_t> stops(0);
+
+    auto on_start = [&]() { ++starts; };
+    auto on_stop = [&]() { ++stops; };
+
+    auto exec = executor_example::make_executor_with_thread_hooks(
+        hpx::parallel::execution::par.executor(), on_start, on_stop);
+
+    hpx::parallel::for_loop(hpx::parallel::execution::par.on(exec), 0, v.size(),
+        [](std::size_t i) {});
+
+    std::cout << "Executed " << starts.load() << " starts and " << stops.load()
+              << " stops\n";
```

Comment:
```suggestion
              << " stops\n";
              HPX_ASSERT(starts > 0);
              HPX_ASSERT(stops > 0);
```
just as a sanity check?

Reply:
Ok, I'll add those tests, but will do so a `HPX_TEST_NEQ(starts, std::size_t(0))`.

Reply:
Hold on... This is an example, not a test. So, `HPX_ASSERT` it is.